src/js/components/FileInput/FileInput.js

```diff
@@ -221,198 +296,165 @@ const FileInput = forwardRef(
-          onChange={(event) => {
-            event.persist();
-            const fileList = event.target.files;
-            const nextFiles = multiple ? [...files] : [];
-            for (let i = 0; i < fileList.length; i += 1) {
-              // avoid duplicates
-              const existing =
-                nextFiles.filter(
-                  (file) =>
-                    file.name === fileList[i].name &&
-                    file.size === fileList[i].size,
-                ).length > 0;
-              if (!existing) nextFiles.push(fileList[i]);
-            }
-            setFiles(nextFiles);
-            setDragOver(false);
-            if (onChange) onChange(event, { files: nextFiles });
-          }}
-        />
-      </ContentsBox>
+            ))}
+          <StyledFileInput
+            ref={inputRef}
+            type="file"
+            id={id}
+            name={name}
+            multiple={multiple}
+            disabled={disabled}
+            plain
+            rightOffset={rightOffset}
+            {...rest}
+            onDragOver={() => setDragOver(true)}
+            onDragLeave={() => setDragOver(false)}
+            onChange={(event) => {
+              event.persist();
+              const fileList = event.target.files;
+              let filesToAdd;
+              if (!files.length) {
+                if (multiple && max && fileList.length >= max) {
+                  const newFileList = [];
+                  for (let i = 0; i < max; i += 1) {
+                    newFileList.push(fileList[i]);
+                  }
+                  filesToAdd = newFileList;
+                } else {
+                  filesToAdd = fileList;
+                }
+              } else if (
+                multiple &&
+                max &&
+                fileList.length >= max - files.length
+              ) {
+                const newFileList = [];
+                for (let i = 0; i < max - files.length; i += 1) {
+                  newFileList.push(fileList[i]);
+                }
+                filesToAdd = newFileList;
+              } else {
+                filesToAdd = fileList;
+              }
+              const nextFiles = multiple ? [...files] : [];
+              for (let i = 0; i < filesToAdd.length; i += 1) {
+                // avoid duplicates
+                const existing =
+                  nextFiles.filter(
+                    (file) =>
+                      file.name === fileList[i].name &&
+                      file.size === fileList[i].size,
+                  ).length > 0;
+                if (!existing) nextFiles.push(filesToAdd[i]);
+              }
+              setFiles(nextFiles);
+              setDragOver(false);
+              if (onChange) onChange(event, { files: nextFiles });
+            }}
+          />
+        </ContentsBox>
+        {max && (
+          <Message {...theme.fileInput.message.maxFile}>
+            {messages?.maxFile
```

Comment:
this needs to be done with `format` like line 382 and others. This message also needs to be added to default.json.

Reply:
Can you suggest how can I add a default maxFile in `default.json`, since I also required `max` prop value ?

Reply:
I'm talking about our default messages file for localization. I think you're thinking of the default properties values.
In [src/js/languages/default.json](https://github.com/grommet/grommet/blob/master/src/js/languages/default.json) between lines 10 and 11 add
```
"maxFile": "Attach a maximum of {max} files only.",
```
Note there is no $ in this one.

Then here you can do:
```jsx
<Message {...theme.fileInput.message}>
  {format({
    id: 'fileInput.maxFile',
    messages,
    values: { max },
   })}
</Message>
```

This will get the text from default.json (or let it be overridden by a `maxFiles` in the `messages` property if present) and replace the `{max}` token in the text with the value passed in for `max`.